lib/tasks/integration.rake

```diff
@@ -0,0 +1,95 @@
+INTEGRATION_TMP_DIR     = Rails.root.join('tmp', 'integration')
+INTEGRATION_FOREMAN_PID = INTEGRATION_TMP_DIR.join('server.pid')
+INTEGRATION_FOREMAN_LOG = Rails.root.join('log', 'integration.log')
+INTEGRATION_PROCFILE    = INTEGRATION_TMP_DIR.join('Procfile')
+RUN_SINGLE_WORKER_BIN   = Rails.root.join('lib', 'workers', 'bin', 'run_single_worker.rb')
+
+namespace :integration do
+  desc "Seed the database and configure assets for integration tests"
+  task :seed => [:env, "test:vmdb:setup", "evm:foreman:seed", :compile_assets]
+
+  desc "Setup the database for integration tests"
+  task :setup => [:db_setup, "evm:foreman:setup"] do
+    mkdir_p INTEGRATION_TMP_DIR
+    rm_f    INTEGRATION_PROCFILE
+    touch   INTEGRATION_PROCFILE
+  end
+
+  desc "Run a foreman server in the background for integration tests"
+  task :run_server => [:setup] do
+    pid = File.read(INTEGRATION_FOREMAN_PID).to_i if File.exist?(INTEGRATION_FOREMAN_PID)
+
+    if pid && (Process.getpgid(pid) rescue nil).present?
+      warn "foreman process already running"
+    else
+      foreman_cmd = %W[
+        foreman start --port=#{ENV['FOREMAN_PORT'] || '3000'}
+                      --root=#{Rails.root}
+                      --procfile=#{INTEGRATION_PROCFILE}
+      ].join(" ")
+
+      pid = Process.spawn(foreman_cmd, [:out, :err] => [INTEGRATION_FOREMAN_LOG, "a"])
+      Process.detach(pid)
+      File.write(INTEGRATION_FOREMAN_PID, pid)
+    end
+  end
+
+  desc "Stop server for integration tests"
+  task :stop_server do
+    if File.exist?(INTEGRATION_FOREMAN_PID)
+      Process.kill("TERM", File.read(INTEGRATION_FOREMAN_PID).to_i)
+      rm INTEGRATION_FOREMAN_PID
+    else
+      warn "PID file for server (#{INTEGRATION_FOREMAN_PID}) doesn't exist... moving on"
+    end
+  end
+
+  desc "With a UI worker for integration tests"
+  task :with_ui => :setup do
+    ui_config = "ui: env PORT=$PORT ruby #{RUN_SINGLE_WORKER_BIN} MiqUiWorker\n"
+    File.write(INTEGRATION_PROCFILE, ui_config, :mode => "a")
+  end
+
+  # Check if a UI worker is running
+  task :ui_ready => :run_server do
```

Comment:
I am going to head off any "just use `wait-on`" with a bit of a rational for going this route:

- I hate excessive shelling out (pedantic... I know, but relevant for later bullets)
- I wanted this to be a rake task to be able to be used as a prerequisite for other tasks
- I wanted to be able to control the timeout
- I wanted this to have a nice user interface that users can query for, and not have to remember
  - In other words: `rake -T integration`

I have already been told/reminded about the use of `wait-on` a few times now by LJ, as very succinctly summed up as to "why" by this comment:

https://github.com/ManageIQ/manageiq-ui-classic/pull/6683#pullrequestreview-421082542

And while if this was a **only** a script, I would agree this is a perfect solution to use (even if the library is [written in js](https://github.com/jeffbski/wait-on)...).  However, the line being referenced:

```bash
bundle exec rails server > /dev/null & wait-on http://localhost:3000
```

Is dense, and there is a lot of "bash/shell-isms" being done (redirecting output, backgrounding a process, and executing a follow up) that I would never want to tell someone to run themselves unless I knew they were competent in their bash-foo \*.  Also, if I were to use this as a `system()` call, then the code would then me rescuing errors from `wait-on`, and probably adding just as much wrapper code as what is here when just using `net/http`.

* * *

So, I think with all of that, I disagree the suggestion to use `wait-on` and thing this is a completely reasonable solution and I don't think it saves us anything by using it.

\* I say this because there are a lot of developers out there who are more IDE focused, and a command like that might not be easy to parse, or realize that an un-monitored background process has been started.  I myself tend to look at "what travis does" when I am trying to reproduced an error from CI, and I prefer when an reproducer setup is as few commands as possible so CI and dev are as close to parity as possible as then it is easier for devs to debug locally when needed.

Reply:
This is a bit of a follow up to the above, but I figured I should at least give
`wait-on` a valid test before completely writing it off.

More details below in a `<details>` block since it is a bit wordy...

<details>
<summary>Details for "Why I still don't like `wait-on`" after testing locally</summary>

I actually started looking through the code a bit to understand how it works, and see if anything I was doing with my ruby script could have been improved to make it a bit more streamlined.

Turns out it's defaults are actually pretty terrible and "webservice unfriendly", as one of the most alarming things I noticed was the default for the `-s`/`--simultaneous` option is `Infinity`... and knew immediately that this would not be a good thing.  What that ends up meaning is that it will make as many HTTP connections as possible, syncronosly, while trying connect on all of them.

So using the previously mentioned inline script as an example:

```console
$ bundle exec rails server > /dev/null & wait-on http://localhost:3000
```

means that `bin/rails s` will try and boot in the background, and then `wait-on` will immediately start pounding it with connections.  I simulated this is a bit more measurable fashion by doing the above in a two step process:

1. Opened a new console and ran `bin/rails s` until I saw the following:
   
   ```console
   $ bin/rails s
   => Booting Puma
   => Rails 5.2.4.3 application starting in development
   => Run `rails server -h` for more startup options
   ** Using session_store: ActionDispatch::Session::MemCacheStore
   ** ManageIQ master, codename: Kasparov
   Puma starting in single mode...
   * Version 4.3.5 (ruby 2.6.3-p62), codename: Mysterious Traveller
   * Min threads: 5, max threads: 5
   * Environment: development
   * Listening on tcp://127.0.0.1:3000
   * Listening on tcp://[::1]:3000
   Use Ctrl-C to stop
   ```
   
2. Ran `time` + `wait-on` to see how long the first request took return.

This gave enough time for the socket to already be open, which is something we do have rescue for in my ruby implementation and aren't testing for here.  `wait-on` would most likely handle this, so that isn't a deficiency in it's implementation, but it wasn't something I wanted to try find a way to test in a reproducible manner.

Anyway, this is the result of the first `time wait-on ...` call after the server was ready:

```console
$ time wait-on http://localhost:3000

real    0m47.685s
user    0m0.481s
sys     0m0.091s
```

and a first snippet of log output from kinda shows why this is a problem (and why it took almost 50s for `wait-on` to return for the first call):

```
...
Puma starting in single mode...
* Version 4.3.5 (ruby 2.6.3-p62), codename: Mysterious Traveller
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://127.0.0.1:3000
* Listening on tcp://[::1]:3000
Use Ctrl-C to stop
[----] I, [2020-07-10T19:08:08.771532 #52344:3fdfd3e14e28]  INFO -- : Started HEAD "/" for 127.0.0.1 at 2020-07-10 19:08:08 -0500
[----] I, [2020-07-10T19:08:09.502617 #52344:3fdfd3e14ae0]  INFO -- : Started HEAD "/" for 127.0.0.1 at 2020-07-10 19:08:09 -0500
[----] I, [2020-07-10T19:08:09.503287 #52344:3fdfd3e14770]  INFO -- : Started HEAD "/" for 127.0.0.1 at 2020-07-10 19:08:09 -0500
[----] I, [2020-07-10T19:08:10.913880 #52344:3fdfd3e14e28]  INFO -- : Processing by DashboardController#login as HTML
[----] I, [2020-07-10T19:08:10.914256 #52344:3fdfd3e14770]  INFO -- : Processing by DashboardController#login as HTML
[----] I, [2020-07-10T19:08:10.915105 #52344:3fdfd3e146bc]  INFO -- : Started HEAD "/" for 127.0.0.1 at 2020-07-10 19:08:10 -0500
[----] I, [2020-07-10T19:08:10.915329 #52344:3fdfd3e14fa4]  INFO -- : Started HEAD "/" for 127.0.0.1 at 2020-07-10 19:08:10 -0500
[----] I, [2020-07-10T19:08:10.914642 #52344:3fdfd3e14ae0]  INFO -- : Processing by DashboardController#login as HTML
```

As you can see, there are a bunch `HEAD` requests queued up, and what ends up happening is it queues up as many `HEAD` requests as feasibly possible, and (I think) waits for them all to return and at least 1 returns a 200 status.  I snipped this log output to make it digestible, but the server was actually still responding to all of the queued up requests **AFTER** `wait-on` returned, which also isn't good.

Now, I know what you might be thinking:

> "Nick... why not just use the `--simultaneous` CLI flag?"

I did, but I think a comparison of that v.s. `curl /api/ping` can show why our app doesn't really respond to the defaults favorably:

### Using `wait-on`

**Console 1**

```
$ bin/rails s
=> Booting Puma
=> Rails 5.2.4.3 application starting in development
=> Run `rails server -h` for more startup options
** Using session_store: ActionDispatch::Session::MemCacheStore
** ManageIQ master, codename: Kasparov
Puma starting in single mode...
* Version 4.3.5 (ruby 2.6.3-p62), codename: Mysterious Traveller
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://127.0.0.1:3000
* Listening on tcp://[::1]:3000
Use Ctrl-C to stop
[----] I, [2020-07-10T19:12:05.310292 #54775:3fc6404e16f0]  INFO -- : Started HEAD "/" for 127.0.0.1 at 2020-07-10 19:12:05 -0500
[----] I, [2020-07-10T19:12:07.278520 #54775:3fc6404e16f0]  INFO -- : Processing by DashboardController#login as HTML
[----] D, [2020-07-10T19:12:07.290016 #54775:3fc6404e16f0] DEBUG -- : PostgreSQLAdapter#log_after_checkout, connection_pool: size: 5, connections: 1, in use: 1, waiting_in_queue: 0
[----] D, [2020-07-10T19:12:07.305113 #54775:3fc6404e16f0] DEBUG -- :   Tenant Load (0.6ms)  SELECT  "tenants".* FROM "tenants" WHERE "tenants"."id" BETWEEN $1 AND $2 AND "tenants"."ancestry" IS NULL ORDER BY "tenants"."id" ASC LIMIT $3  [["id", 10000000000000], ["id", 10999999999999], ["LIMIT", 1]]
[----] D, [2020-07-10T19:12:07.322051 #54775:3fc6404e16f0] DEBUG -- :   Tenant Inst Including Associations (16.6ms - 1rows)
[----] D, [2020-07-10T19:12:07.326988 #54775:3fc6404e16f0] DEBUG -- :   MiqServer Load (0.6ms)  SELECT  "miq_servers".* FROM "miq_servers" WHERE "miq_servers"."guid" = $1 LIMIT $2  [["guid", "c539402d-fc7a-4f1e-9fbc-1c026e16b14c"], ["LIMIT", 1]]
[----] D, [2020-07-10T19:12:07.327462 #54775:3fc6404e16f0] DEBUG -- :   MiqServer Inst Including Associations (0.1ms - 1rows)
[----] I, [2020-07-10T19:12:07.332172 #54775:3fc6404e16f0]  INFO -- :   Rendering /Users/nicklamuro/code/redhat/manageiq-ui-classic/app/views/dashboard/login.html.haml within layouts/login
[----] D, [2020-07-10T19:12:14.329410 #54775:3fc6404e16f0] DEBUG -- :   CACHE MiqServer Load (0.0ms)  SELECT  "miq_servers".* FROM "miq_servers" WHERE "miq_servers"."guid" = $1 LIMIT $2  [["guid", "c539402d-fc7a-4f1e-9fbc-1c026e16b14c"], ["LIMIT", 1]]
[----] D, [2020-07-10T19:12:14.329703 #54775:3fc6404e16f0] DEBUG -- :   MiqServer Inst Including Associations (0.1ms - 1rows)
[----] I, [2020-07-10T19:12:14.329875 #54775:3fc6404e16f0]  INFO -- :   Rendered /Users/nicklamuro/code/redhat/manageiq-ui-classic/app/views/layouts/_spinner.html.haml (2.9ms)
[----] I, [2020-07-10T19:12:14.335377 #54775:3fc6404e16f0]  INFO -- :   Rendered /Users/nicklamuro/code/redhat/manageiq-ui-classic/app/views/layouts/_flash_msg.html.haml (4.5ms)
[----] I, [2020-07-10T19:12:14.340277 #54775:3fc6404e16f0]  INFO -- :   Rendered /Users/nicklamuro/code/redhat/manageiq-ui-classic/app/views/dashboard/_login_more.html.haml (3.6ms)
[----] D, [2020-07-10T19:12:14.531813 #54775:3fc6404e16f0] DEBUG -- :   Zone Load (0.5ms)  SELECT  "zones".* FROM "zones" WHERE "zones"."id" = $1 LIMIT $2  [["id", 10000000000001], ["LIMIT", 1]]
[----] D, [2020-07-10T19:12:14.532210 #54775:3fc6404e16f0] DEBUG -- :   Zone Inst Including Associations (0.1ms - 1rows)
[----] I, [2020-07-10T19:12:14.532646 #54775:3fc6404e16f0]  INFO -- :   Rendered /Users/nicklamuro/code/redhat/manageiq-ui-classic/app/views/dashboard/login.html.haml within layouts/login (7200.3ms)
[----] I, [2020-07-10T19:12:14.538060 #54775:3fc6404e16f0]  INFO -- :   Rendered /Users/nicklamuro/code/redhat/manageiq-ui-classic/app/views/layouts/_doctype.html.haml (0.8ms)
[----] I, [2020-07-10T19:12:14.979551 #54775:3fc6404e16f0]  INFO -- :   Rendered /Users/nicklamuro/code/redhat/manageiq-ui-classic/app/views/layouts/_i18n_js.html.haml (1.9ms)
[----] D, [2020-07-10T19:12:14.980190 #54775:3fc6404e16f0] DEBUG -- :   CACHE MiqServer Load (0.0ms)  SELECT  "miq_servers".* FROM "miq_servers" WHERE "miq_servers"."guid" = $1 LIMIT $2  [["guid", "c539402d-fc7a-4f1e-9fbc-1c026e16b14c"], ["LIMIT", 1]]
[----] D, [2020-07-10T19:12:14.980631 #54775:3fc6404e16f0] DEBUG -- :   MiqServer Inst Including Associations (0.2ms - 1rows)
Session:         Hash of Size 3731, Elements 12
=================================
[----] I, [2020-07-10T19:12:14.981730 #54775:3fc6404e16f0]  INFO -- : Completed 200 OK in 7703ms (Views: 7652.8ms | ActiveRecord: 18.0ms)


[----] D, [2020-07-10T19:12:14.983636 #54775:3fc6404e16f0] DEBUG -- : PostgreSQLAdapter#log_after_checkin, connection_pool: size: 5, connections: 1, in use: 0, waiting_in_queue: 0
```

**Console 2**

```console
$ time wait-on -s 1 http://localhost:3000

real    0m9.949s
user    0m0.320s
sys     0m0.048s
```


### Using `curl`

**Console 1**

```
$ bin/rails s
=> Booting Puma
=> Rails 5.2.4.3 application starting in development
=> Run `rails server -h` for more startup options
** Using session_store: ActionDispatch::Session::MemCacheStore
** ManageIQ master, codename: Kasparov
Puma starting in single mode...
* Version 4.3.5 (ruby 2.6.3-p62), codename: Mysterious Traveller
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://127.0.0.1:3000
* Listening on tcp://[::1]:3000
Use Ctrl-C to stop
[----] I, [2020-07-10T19:13:01.639740 #55310:3fdbe26882a8]  INFO -- : Started GET "/api/ping" for ::1 at 2020-07-10 19:13:01 -0500
[----] I, [2020-07-10T19:13:02.392739 #55310:3fdbe26882a8]  INFO -- : Processing by Api::PingController#index as JSON
[----] I, [2020-07-10T19:13:02.394123 #55310:3fdbe26882a8]  INFO -- : Completed 200 OK in 1ms (Views: 0.9ms)
```

**Console 2**

```console
$ time curl http://localhost:3000/api/ping
pong
real    0m0.774s
user    0m0.005s
sys     0m0.004s
```

Now this isn't a completely fair comparison, since `/api/ping` is a much lighter weight route and can load up quicker on first boot.  That said, there was still some overhead (at `~200ms`, sometimes more) with `wait-on` that I wasn't seeing with `curl`:

```console
# Between each of these calls, I restarted the server
$ time wait-on -s 1 http://localhost:3000/api/ping

real    0m0.914s
user    0m0.274s
sys     0m0.042s
$ time wait-on -s 1 http://localhost:3000/api/ping

real    0m1.021s
user    0m0.279s
sys     0m0.046s
$ time curl http://localhost:3000/api/ping
pong
real    0m0.762s
user    0m0.005s
sys     0m0.006s
```

The only time I have record of `curl` being slower was when the request by `curl` was done against the same server process that was processing the default `wait-on http://localhost:3000` call, and I am pretty sure it was still working on the queued up processes as a result.
</details>

Reply:
So, if I'm trying to follow this correctly...

wait-on by default uses constant HEAD calls during boot which ends up flooding the first call, which in turn ends up doing a lot of UI work.  But then later you found that wait-on can use the api/ping, and the time difference is now only 200ms (on an already subsecond call)?  In that latter api/ping case does it still flood the pipeline with useless head calls?  Does it even matter at that point?  I'd argue maintaining our own polling code has it's own cost, given there's already one that already comes with cypress.

> $ bundle exec rails server > /dev/null & wait-on http://localhost:3000
> ...
> and a command like that might not be easy to parse, or realize that an un-monitored background process has been started

I'd argue that's incredibly easy to read, and very clear that a Rails server is being started separate and then waiting for it to be ready on port 3000.  We probably should also shut down the Rails server when the testing is complete (unless maybe CYPRESS_DEV is set, or some other option).

Reply:
Note, I'm not against using the custom polling mechanism, just questioning the above since it seemed more of a journey that landed on something that's ultimately not as bad as originally described.  If the Rails request queue is still flooded though, and the argument about "wrapping `system()` calls for errors anyway" still applies, then I am fine with it.

Reply:
> So, if I'm trying to follow this correctly...

Sorry this got long and was a bit of a "stream of conscious" when doing it, which is major reason why I put the second bit in a `<details>`... unfortunately my response isn't much better...

> But then later you found that wait-on can use the api/ping...

<details>
<summary>answer...</summary>

It can, but also keep in mind that when I was testing `/api/ping`, I was already using `-s 1` (aka `--simultaneous 1` ) to avoid the default of `Infinity` for that.  So it probably would spam the server in the same way with a few requests as well, and definitely would if used like it was in the previous use case:

```console
$ bundle exec rails server > /dev/null & wait-on http://localhost:3000/api/ping
```

Where the server isn't even accepting requests for the first couple of seconds.

Also, on a similar note, `-t, --timeout` also defaults to `Infinity`, so that means that what was originally there would hang if the server didn't start up properly.  And even if we include that flag, the error that is eventually returned isn't very helpful....

```
$ wait-on -s 1 -t 5000 http://localhost:3000
Error: Timeout
    at MergeMapSubscriber.project (/usr/local/lib/node_modules/wait-on/lib/wait-on.js:128:74)
    at MergeMapSubscriber._tryNext (/usr/local/lib/node_modules/wait-on/node_modules/rxjs/internal/operators/mergeMap.js:69:27)
    at MergeMapSubscriber._next (/usr/local/lib/node_modules/wait-on/node_modules/rxjs/internal/operators/mergeMap.js:59:18)
    at MergeMapSubscriber.Subscriber.next (/usr/local/lib/node_modules/wait-on/node_modules/rxjs/internal/Subscriber.js:66:18)
    at AsyncAction.dispatch [as work] (/usr/local/lib/node_modules/wait-on/node_modules/rxjs/internal/observable/timer.js:31:16)
    at AsyncAction._execute (/usr/local/lib/node_modules/wait-on/node_modules/rxjs/internal/scheduler/AsyncAction.js:71:18)
    at AsyncAction.execute (/usr/local/lib/node_modules/wait-on/node_modules/rxjs/internal/scheduler/AsyncAction.js:59:26)
    at AsyncScheduler.flush (/usr/local/lib/node_modules/wait-on/node_modules/rxjs/internal/scheduler/AsyncScheduler.js:52:32)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)
```
</details>

> I'd argue that's incredibly easy to read, and very clear that a Rails server is being started separate and then waiting for it to be ready on port 3000.

<details>
<summary>answer...</summary>

Maybe from a "high level conceptual" stand point of "start one thing and wait for it to start", sure, but not from a true understanding standpoint.

I googled for "bash &" to clarify myself the specifics, and this is the first thing that came up:

https://stackoverflow.com/questions/26770568/vs-with-the-test-command-in-bash/26770612

and kind of gets to the point I am trying to make.  Many people are confused at the "not so subtle" differences of `&` v.s. `&&` in shell programming, and might not understand that the `&` is critical to background the process and mistake it for a `&&` ("do this, then that").

In reality, the above sample should be more easily understood as:

```console
$ bundle exec rails server > /dev/null &  # background `rails server` and redirect output
$ wait-on http://localhost:3000/api/ping
```

which even then, my comment has a lot of extra explaining to do.

</details>

* * *

I think a point that I missed in addition to all of my previous "reasons" (and I just added it) was:

> - I don't want `wait-on` as an extra dependency to use these tasks

And I think that is important point to make.  I think adding it is another bit of maintenance in it of itself, and when I am trying to make the tasks here as approachable as possible, adding an extra hurdle in doing that.

If we did require it, then how would we have them install it?  Would they do their own `npm install -g wait-on`, or have it included in the `manageiq-ui-classic` `package.json`?  If we have them maintain it themselves, then do we have to include documentation?  Do we make sure it is a specific version?  How do we sync that up with what is installed by default on `Travis`?

If it is included as part of the `package.json`, how do we then execute it since it isn't global?  Most likely it would require a helper method to find the path for that on a given machine, so will there need to be specifics for `Travis` there as well?  Do we use a config file for the `-s`/`-t`/etc. or just write them inline?  What do we do when it errors?

And then debugging.  For me just to write the ["lengthy explanation above"](https://github.com/ManageIQ/manageiq/pull/20350#discussion_r453133445), I had to dig not only into the code base of `wait-on`, but also [`rxjs`](https://rxjs-dev.firebaseapp.com/api) since that is a dependency of `wait-on` and figure out how "Observables" work.  I question how much, if not more, would be required to debug issues when the server is failing to start on `travis`, and I fail to see how this isn't just as much of a maintenance burden in a different way as just writing 5 lines of stdlib Ruby.

Reply:
Just to quickly (seriously) follow up on this:

> Note, I'm not against using the custom polling mechanism, just questioning the above since it seemed more of a journey that landed on something that's ultimately not as bad as originally described.

"The journey" was more why I didn't see any benefits for our use case to use `wait-on`, and if any, and making note of potential pitfalls/maintenance/tech-debt that library could also bring that isn't immediately obvious.  That said, it was more "additional info" versus "the only argument" I was making, which is why I left most of it in a `<details>` tag.

Just like with ["Ruby bin scripts"](https://github.com/ManageIQ/manageiq-style/pull/5#pullrequestreview-435321574), I prefer the logic in a script instead of shelling out, specifically if I want it to run in multiple places.  I try to reserve shell scripts for "automating (sequential) processes", not "automating (branching) logic".  The `net/http` approach has some conditional logic to it, and I think `wait-on` approach arguably does as well, and in my opinion trying to port that logic over to a dev environment is much easier to do within the confines of the language we are already "automating logic" in ("Ruby `rake` tasks"), instead of trying to call out to an external tool to do so.

Reply:
Going to resolve this not because it is necessarily "resolved", but because it takes up a lot of space (my fault).  If anyone wishes to comment on or disagree with the arguments above, feel free to "unresolve" and do so.